### PR TITLE
Feat/kandel

### PR DESCRIPTION
**Pull request by lnist**
KandelStrategies is the entry point.
The farm can retrieve existing instances.
The seeder can sow instances.
The instance can manage an instance.

Note: There are still outstanding issues
1) Integrate with final mangrove-core
2) Implement missing functionality
3) Better chunking

Example script which can be extended to a tutorial:
/*

UI or DOC disclaimer:
If using liquidity sharing then UI may not show correct values.
If manually changing the router then UI may not show correct values.
If deploying kandel outside seeder then UI does not show it.

*/
```typescript
// Load the RPC_URL and PRIVATE_KEY from .env file into process.env
// This script assumes RPC_URL points to your access point and PRIVATE_KEY contains private key from which one wishes to post offers
var parsed = require("dotenv").config();
// Import the Mangrove API
const {
  Mangrove,
  ethers,
  KandelStrategies,
} = require("@mangrovedao/mangrove.js");

// Create a wallet with a provider to interact with the chain.
const provider = new ethers.providers.WebSocketProvider(process.env.LOCAL_URL); // For local chain use
const wallet = new ethers.Wallet(process.env.PRIVATE_KEY, provider); // Use either own account or if on local chain use an anvil account

// Connect the API to Mangrove
const mgv = await Mangrove.connect({ signer: wallet });

// Get helper for working with strategies
const kandelStrategies = new KandelStrategies({ mgv });

// Choose a market
const market = await mgv.market({ base: "WETH", quote: "USDC" });

// Get balances to see what we have to work with
await market.base.balanceOf(await mgv.signer.getAddress());
await market.quote.balanceOf(await mgv.signer.getAddress());

// Get a distribution generator
let generator = kandelStrategies.generator(market);

// Calculate a candidate distribution (should be based on minimum gives/wants)
let distribution = generator.calculateDistribution({
  priceParams: { minPrice: 900, maxPrice: 1100, ratio: 1.01 },
  midPrice: 1000,
  initialAskGives: 1,
  initialBidGives: 1000,
});
let offeredVolume = generator.getOfferedVolumeForDistribution(distribution);

// Recalculate the distribution based on what we want to offer.
let distribution2 = generator.recalculateDistributionFromAvailable({
  distribution,
  availableBase: 1,
  availableQuote: 1000,
});
let offeredVolume2 = generator.getOfferedVolumeForDistribution(distribution2);

// Deploy a kandel
const { kandelPromise } = await kandelStrategies.seeder.sow({
  onAave: false,
  market,
  liquiditySharing: false,
  gaspriceFactor: 10,
});
const kandel = await kandelPromise;

// Approve kandel to use our funds
let txs = await kandel.approve();
await Promise.all(txs.map((x) => x.wait()));

//TODO a bit weird to have to pass length here... and ratio.
// Populate the kandel according to our desired distribution
// Note: we need to pass in the pricePoints since this could be a subset of the entire distribution
// Note: we need to pass in the ratio so that kandel knows what the distribution was calculated based on.
txs = await kandel.populate({
  distribution: distribution2,
  parameters: {
    ratio: 1.01,
    compoundRateBase: 1.0,
    compoundRateQuote: 1.0,
    pricePoints: distribution2.length,
    spread: 1,
  },
  depositBaseAmount: 0,
  depositQuoteAmount: 0,
});

await Promise.all(txs.map((x) => x.wait()));
```

### CODE REVIEW

File: packages/mangrove.js/src/kandel/kandelInstance.ts [hunk 1]
```diff
@@ -0,0 +1,805 @@
...
+    }
+    return current;
+  }
+
+  private offerTypeToUint(offerType: Market.BA): number {
+    return offerType == "bids" ? 0 : 1;
+  }
+
+  private UintToOfferType(offerType: number): Market.BA {
+    return offerType == 0 ? "bids" : "asks";
```
[the remark this answers is not in the document]
Comment (lnist) [reply packages/mangrove.js/src/kandel/kandelInstance.ts]: yes (for now :D )

File: packages/mangrove.js/src/kandel/kandelInstance.ts [hunk 2]
```diff
@@ -0,0 +1,805 @@
...
+      overrides
+    );
+  }
+
+  async getRawDistributionChunks(params: {
+    distribution: Distribution;
+    maxOffersInChunk?: number;
+  }) {
+    this.generator.distributionHelper.sortByIndex(params.distribution);
+    this.verifyDistribution(params.distribution);
```
[the remark this answers is not in the document]
Comment (lnist) [reply packages/mangrove.js/src/kandel/kandelInstance.ts:530]: sortByIndex is already returning the sorted (but it also sorts in-place as that is what javascript does)
verify throws with various errors depending on the error case so it should not return.

File: packages/mangrove.js/src/kandel/kandelPriceCalculation.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+export type PriceDistributionParams = {
+  minPrice?: Bigish;
+  maxPrice?: Bigish;
+  ratio?: Bigish;
+  pricePoints?: number;
+};
+
+/** @title Helper for calculating details about about a Kandel instance. */
+class KandelPriceCalculation {
+  public calculatePrices(params: PriceDistributionParams) {
```
[the remark this answers is not in the document]
Comment (lnist) [reply packages/mangrove.js/src/kandel/kandelPriceCalculation.ts:29]: Fixed

File: packages/mangrove.js/src/kandel/kandelStatus.ts [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+      return { i, diff: midPrice.minus(x).abs() };
+    });
+    diffs.sort((a: { diff: Big }, b: { diff: Big }) =>
+      a.diff.gt(b.diff) ? 1 : b.diff.gt(a.diff) ? -1 : 0
+    );
+
+    return diffs[0].i;
+  }
+
+  public getOfferStatuses(
```
[the remark this answers is not in the document]
Comment (lnist) [reply packages/mangrove.js/src/kandel/kandelStatus.ts:117]: fixed

File: packages/mangrove.js/src/util/test/mochaHooks.ts [hunk 5]
```diff
@@ -97,6 +97,7 @@ export const mochaHooks = {
         if (newProxy.cancelAll) {
           console.log("mochaHooks: Got request after cancelled!");
           session.destroy();
+          //console.dir(data.toString());
```
[the remark this answers is not in the document]
Comment (lnist) [reply packages/mangrove.js/src/util/test/mochaHooks.ts]: nope - but I added more info.

File: packages/mangrove.js/test/integration/kandel.integration.test.ts [hunk 6]
```diff
@@ -0,0 +1,973 @@
...
+        "should use specified gasprice and multiplier."
+      );
+    });
+  });
+
+  describe("farm", async function () {
+    let farm: KandelFarm;
+    let defaultOwner: string;
+
+    beforeEach(async function () {
```
[the remark this answers is not in the document]
Comment (lnist) [reply packages/mangrove.js/test/integration/kandel.integration.test.ts:141]: Nope - anvil is reverted to earlier snapshot.

File: packages/mangrove.js/src/kandel/kandelDistributionHelper.ts [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+          .round(
+            offerType == "asks" ? this.baseDecimals : this.quoteDecimals,
+            Big.roundDown
+          )
+      : Big(0);
+  }
+
+  public calculateConstantOutboundPerOffer(
+    distribution: Distribution,
+    availableBase: Big,
```
[the remark this answers is not in the document]
Comment (lnist) [reply packages/mangrove.js/src/kandel/kandelDistributionHelper.ts:60]: Good idea - will put it on backlog as a consideration for now.

File: packages/mangrove.js/src/kandel/kandelFarm.ts [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+      "AaveKandelSeeder",
+      this.mgv.network.name
+    );
+    this.aaveKandelSeeder = typechain.AaveKandelSeeder__factory.connect(
+      aaveKandelSeederAddress,
+      this.mgv.signer
+    );
+  }
+
+  getTokenAndAddress(address: string) {
```
[the remark this answers is not in the document]
Comment (lnist) [reply packages/mangrove.js/src/kandel/kandelFarm.ts:42]: Moved to mgv.

File: packages/mangrove.js/src/kandel/kandelDistributionHelper.ts [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        ? {
+            index,
+            base: bidGives.div(p).round(this.baseDecimals, Big.roundHalfUp),
+            quote: bidGives,
+            offerType: "bids" as Market.BA,
+          }
+        : {
+            index,
+            base: askGives,
+            quote: askGives.mul(p).round(this.quoteDecimals, Big.roundHalfUp),
```
[the remark this answers is not in the document]
Comment (lnist) [reply packages/mangrove.js/src/kandel/kandelDistributionHelper.ts]: Pulled out in functions.

File: packages/mangrove.js/src/kandel/kandelDistributionHelper.ts [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+  /** Gets the index of the first ask in the distribution
+   * @param distribution The distribution to get the index from.
+   * @returns The index of the first ask in the distribution.
+   */
+  public getFirstAskIndex(distribution: Distribution) {
+    return (
+      distribution.find((x) => x.offerType == "asks")?.index ??
+      distribution.length
+    );
+  }
```
Comment (peterMangrove) [inline packages/mangrove.js/src/kandel/kandelDistributionHelper.ts:205]: Is it intuitive that the default value is the length of the distribution? Should the docs, explain the default value?

Comment (lnist) [reply packages/mangrove.js/src/kandel/kandelDistributionHelper.ts:205]: Comment updated

**Pull request merged**